tests/test_pandas.py

```diff
@@ -0,0 +1,150 @@
+                "CASH",
+                "CREDIT_CARD",
+            ]
+        }
+    )
+
+    actual_df = update_payment_type_as_string_values(df)
+
+    assert_frame_equal(actual_df, expected_df)
+
+
+def test_add_borough_and_zone():
+    trip_df = pd.DataFrame(
+        {
+            "pulocationid": [1, 2, 3, 4, 5],
+            "dolocationid": [1, 2, 3, 4, 5],
+        }
+    )
+
+    zone_df = pd.DataFrame(
+        {
+            "locationid": [1, 2, 3, 4, 5],
+            "borough": ["Manhattan", "Queens", "Bronx", "Brooklyn", "Staten Island"],
+            "zone": ["Midtown", "Astoria", "Riverdale", "Park Slope", "Tottenville"],
+        }
+    )
+
+    actual_df = add_borough_and_zone(trip_df, zone_df, "pulocationid")
+
+    expected_df = pd.DataFrame(
+        {
+            "pulocationid": [1, 2, 3, 4, 5],
+            "dolocationid": [1, 2, 3, 4, 5],
+            "pulocationid_borough": [
+                "Manhattan",
+                "Queens",
+                "Bronx",
+                "Brooklyn",
+                "Staten Island",
+            ],
+            "pulocationid_zone": [
+                "Midtown",
+                "Astoria",
+                "Riverdale",
+                "Park Slope",
+                "Tottenville",
+            ],
+        }
+    )
+
+    assert_frame_equal(actual_df, expected_df)
+
+
+def test_calc_result_most_frequent_three_routes(locations, top3_locations):
+    trip_df = pd.DataFrame(locations)
+
+    actual_df = calc_result_most_frequent_three_routes(trip_df)
+
+    expected_df = pd.DataFrame(top3_locations)
+
+    assert np.array_equal(actual_df.values, expected_df.values)
+
+
+def test_calc_cash_journeys_per_pickup():
+    trip_df = pd.DataFrame(
+        {
+            "payment_type": ["CASH", "CREDIT", "CASH", "NO_CHARGE", "CASH"],
+            "pulocationid_borough": ["1", "2", "1", "3", "1"],
+            "pulocationid_zone": ["1", "1", "1", "3", "4"],
+        }
+    )
+    actual_df = calc_cash_journeys_per_pickup(trip_df)
+
+    expected_df = pd.DataFrame(
+        {
+            "pulocationid_borough": ["1", "1"],
+            "pulocationid_zone": ["1", "4"],
+            "num_cash_journeys": [2, 1],
+        }
+    )
+
+    assert_frame_equal(actual_df, expected_df)
+
+
+def test_calc_highest_tolls_per_route(locations):
+    locations_with_tolls = {
+        **locations,
+        "tolls_amount": [
+            1.0,
+            1.0,
+            1.0,
+            1.0,
+            1.0,
+            1.0,
+            1.0,
+            1.0,
+            1.0,
+            1.0,
+            1.0,
+            1.0,
+            1.0,
+        ],
+    }
+
+    trip_df = pd.DataFrame(locations_with_tolls)
+    actual_df = calc_highest_tolls_per_route(trip_df)
+    actual_df = actual_df[["tolls_amount_sum"]]
+
+    expected_df = pd.DataFrame(
+        {"tolls_amount_sum": [3.0, 3.0, 2.0, 2.0, 1.0, 1.0, 1.0]}
+    )
+
+    assert np.array_equal(actual_df.values, expected_df.values)
```

Comment:
Comparing only the values could lead to bugs caused by differences in the names of columns. I quickly tried to use the DF API and got an index issue. I don't know what causes this, but I'd suggest to exclude index from the check if its unimportant.

Reply:
The issue is, that the Index Types of the compared DataFrames are different.  One time it is a [RangeIndex](https://pandas.pydata.org/docs/reference/api/pandas.RangeIndex.html#pandas.RangeIndex) and in the other case it is a usual Index object. The order of the index rows is also different. 

Using the index in the assert will fail for those two reasons.

What we could do as a workaround is reset_index on both DataFrames before asserting that they are equal. I will implement this workaround.

Reply:
Just as an example: 
![image](https://github.com/johnnylarner/pandas-polars-pyspark/assets/26205656/f5a98cbf-e6cc-4dab-a708-b63f98086f35)

Reply:
Hang on: if you reset the index after the function call, then your test doesn't test how your function works. You can do the reset in the function itself, but doing it after can lead to bugs.  Alternatively, if the index is irrelevant (i'm not really sure what the index does in pandas), then we can change the value of the assert equal kwarg [here](https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_frame_equal.html#:~:text=check_index_typebool%20or%20%7B%E2%80%98equiv%E2%80%99%7D%2C%20default%20%E2%80%98equiv%E2%80%99)

Reply:
What we could do is reset_index on both DataFrames before asserting that they are equal. This asserts that
- values in columns are same
- object is the same (DataFrame)
- index is the same -> of course, as we reset it, this just shows that we have the same number of records (more or less). 

@johnnylarner what do you think about this?

Reply:
@chberreth  I don't think we need to compare the index so you can just set the kwarg to `false`?

Reply:
@johnnylarner the kwags is not working unfortunately

![image](https://github.com/johnnylarner/pandas-polars-pyspark/assets/26205656/bfd72f27-f642-4b24-a4da-401d8ed3ea41)
Any other ideas?

Reply:
Reset in the index in the `calc` function (this is a big down side of pandas imo), try with the `kwarg` ( i think this will just compare the raw values) and then if not just ensure the index is the same for the `expected_df`??

Reply:
So to finish this topic and not overthink it, I'd vote for resetting the index. What do you think?

Reply:
@johnnylarner I reset the index in the calc functions. Correct that is a big downside in pandas. Maybe we should collect all the experiences we made in the repo. What do you think? Maybe we can consolidate them in a later step and use them as a result of the experiement as well.